### ISSUE TITLE
[Discussion] Why would I *not* want to use this?

### ISSUE THREAD

**Issue by serkandurusoy** [SachaG/smartquery#3]
This seems so magical and straightforward, why would I _not_ want to use this for all my mongo cursor publications?

Are there any potential downsides or shortcomings that one could think of?

Few thoughts; one would want
- to combine cursors into a single publication for performance
- joins so prefer something like reywood:publish-composite
- to use observe/observeChanges, transform to modify documents on the fly

**Comment by SachaG** [comment]
Haha :)

Right now the main shortcoming is that subscriptions are getting torn down/recreated every time you click "load more", which obviously is not great… I think this happens because subscriptions get stopped when a reactive context (which helpers are) is invalidated:

> If you call Meteor.subscribe within a reactive computation, for example using Tracker.autorun, the subscription will automatically be cancelled when the computation is invalidated or stopped

Apart from that, I think it could actually be a valid solution to the pub/sub question. It's probably not the most performant thing ever, but it could serve as a good alternative for simpler apps or beginner Meteor devs.

**Comment by serkandurusoy** [comment]
Hmm, is that problem not mitigated by @arunoda's subscription manager?

**Comment by SachaG** [comment]
Well, it would mitigate the issue in theory, but then I'm worried it would open a new sub every time the helper reruns, which is not what we want either…

**Comment by serkandurusoy** [comment]
Hm, we could keep a reference to handles and stop them, but that has its own problems to tackle. Well...

**Comment by SachaG** [comment]
Hmm, maybe… But I feel like there must be away to turn off the auto-stop instead, with [Tracker.nonreactive](http://docs.meteor.com/#/full/tracker_nonreactive)?

**Comment by serkandurusoy** [comment]
Hm, that would be clever, but may lead to potential leakage if some subs are left unstopped. Also, non-unique `id`'s may incur problems. Perhaps if you change the api to not accept subscription `id` but autogenerate them may at least partially simplify the problem.

**Comment by SachaG** [comment]
I would auto-generate them if I could, but I'm not sure how, given that each call to SmartQuery has its own subscription/needs its own id.

**Comment by serkandurusoy** [comment]
Hm, changing the method signature for the publish function from `function (id, selector, options)` to `function (selector, options)`, and setting `id` within the function body using [Random](http://docs.meteor.com/#/full/random) and pushing that to an array outside the function body for tracking purposes would be a start.

BTW, I'm sorry, it seems like I'm just throwing ideas around while I should have created PR's or gists that we could discuss. But I'll do that as soon as I get more time on my hands :)

**Comment by SachaG** [comment]
Yeah but then the id would be different every time the helper is re-run, which is not what we want. And no worries, bouncing ideas around is very helpful too :)
